services/imagery/src/backends/camera-base-backend.js

```diff
@@ -0,0 +1,167 @@
+import request from 'superagent';
+import addProtobuf from 'superagent-protobuf';
+
+import logger from '../common/logger';
+import { imagery, telemetry } from '../messages';
+
+import { removeOrientation, wait } from '../util';
+
+addProtobuf(request);
+
+export default class CameraBaseBackend {
+  /**
+   * Base which takes care of taking images on an interval.
+   *
+   * Takes care of telemetry as well. Requires extension to implement
+   * camera capture behavior.
+   *
+   * Extended classes may implement the acquire() and release(camera)
+   * functions, but must implmenent capture(camera).
+   */
+
+  /** Create a new camera base backend. */
+  constructor(imageStore, interval, telemetryUrl) {
+    this._imageStore = imageStore;
+    this._interval = interval;
+    this._telemetryUrl = telemetryUrl;
+    this._active = false;
+    this._runningMeta = 0;  // Number of running meta tasks.
+    this._endMeta = () => {};  // Used for stop().
+  }
+
+  /** Get the camera and then start the loop in the background. */
+  async start() {
+    this._active = true;
+
+    // Run an acquire function if created on implementor class.
+    if (this.acquire) {
+      this._camera = await this.acquire();
+    } else {
+      this._camera = { };
+    }
+
+    this._loop = this._runLoop();
+  }
+
+  /** Set a flag to kill the camera loop. */
+  async stop() {
+    this._active = false;
+
+    // Wait until the running loop has finished.
+    await this._loop;
+
+    // Wait until all the background tasks have finished.
+    await new Promise((resolve) => {
+      this._endMeta = () => this._runningMeta === 0 ? resolve() : null;
+      this._endMeta();  // In case it's already zero.
+    });
+
+    // Run a release function if created on implementor class.
+    if (this.release) {
+      await this.release();
+    }
+  }
+
+  /** Continuously take photos. */
+  async _runLoop() {
+    while (this._active) {
+      const startTime = Date.now();
+
+      let photo;
+      const metadataPromise = this._getMeta();
+      this._runningMeta++;
+
+      try {
+        photo = await this._takePhoto(this._camera);
+      } catch (err) {
+        const message = err.name + ': ' + err.message;
+        logger.error('Encountered an error in camera loop: ' + message);
+
+        // Update the count whether successful or not.
+        metadataPromise.catch(() => {}).then(() => {
+          this._runningMeta--;
+          this._endMeta();
+        });
+
+        // If we have an error, wait for a little to prevent these
+        // errors from happening too rapidly and then continue;
+        await wait(250);
```

Comment:
If metadata fails to be retrieved, but not the image, it wouldn't wait the 250 ms, would it?

Reply:
Yeah, so if image capture was successful, then it's fine the telemetry request to run in the background if the telemetry request fails. The `metadataPromise` is intentionally not `await`-ed anywhere here.

Reply:
Same for line 96.

Reply:
Actually, metadataPromise should only fail with something really odd happening, since the actual request is `try`/`catch`-ed.

Reply:
I would be wary about letting odd cases slip - I really don't want to take any chances.

Reply:
Even if the telemetry does fail it's `.catch(...)`-ed at line 81 and 98, so it's caught, and it also has a timeout of 1500 ms so it's fine.